File: memory/embeddings.py
"""
memory/embeddings.py

Deterministic Offline Embedding Generator for Roo4u.
Uses multi-scale signed feature hashing with subwords and domain token boosting.
100% offline, zero external models, zero cloud APIs, zero mock dependencies.
"""

import re
import zlib
import hashlib
from typing import List, Tuple, Union
import numpy as np


class OfflineEmbeddingGenerator:
    """
    100% Offline, deterministic multi-scale feature hashing embedding generator.
    Produces L2-normalized float32 vectors in R^D suitable for cosine similarity search.
    """
    def __init__(self, dimension: int = 256):
        self.dimension = int(dimension)
        if self.dimension <= 0:
            raise ValueError(f"Embedding dimension must be positive: {dimension}")

    def _tokenize(self, text: str) -> List[Tuple[str, float]]:
        """
        Tokenizes text into multi-scale lexical, n-gram, and domain features with importance weights.
        """
        if not text or not str(text).strip():
            return [("__EMPTY__", 1.0)]

        tokens: List[Tuple[str, float]] = []
        cleaned = str(text).lower().strip()

        # 1. Extract status codes and special tokens (e.g., 403, 429, 500)
        status_codes = re.findall(r"\b(4\d\d|5\d\d)\b", cleaned)
        for code in status_codes:
            tokens.append((f"status:{code}", 3.0))

        # 2. Extract words and domain/path components
        words = re.findall(r"[a-z0-9_\-\.:#\[\]=\"]+", cleaned)
        for w in words:
            tokens.append((f"w:{w}", 1.5))

        # 3. Extract word bigrams
        for i in range(len(words) - 1):
            tokens.append((f"bi:{words[i]}_{words[i+1]}", 2.0))

        # 4. Extract character n-grams (3-grams and 4-grams) for selector and subword drift tolerance
        for w in words:
            if len(w) >= 3:
                for i in range(len(w) - 2):
                    tokens.append((f"3g:{w[i:i+3]}", 0.5))
            if len(w) >= 4:
                for i in range(len(w) - 3):
                    tokens.append((f"4g:{w[i:i+4]}", 0.5))

        return tokens

    def embed_text(self, text: str) -> np.ndarray:
        """
        Generates a 1D L2-normalized float32 NumPy array of shape (dimension,) for input text.
        Guarantees ||embed_text(text)||_2 == 1.0 within numerical precision.
        """
        vec = np.zeros(self.dimension, dtype=np.float32)
        tokens = self._tokenize(text)

        for token_str, weight in tokens:
            b_str = token_str.encode("utf-8")
            # Bucket index via CRC32
            idx = zlib.crc32(b_str) % self.dimension
            # Sign hash via MD5 high bit for zero-mean projection
            md5_digest = hashlib.md5(b_str).digest()
            sign = 1.0 if (md5_digest[0] & 1) == 0 else -1.0
            vec[idx] += sign * weight

        # L2 normalize
        norm = np.linalg.norm(vec)
        if norm > 1e-12:
            vec = vec / norm
        else:
            vec[0] = 1.0  # Fallback unit vector

        return vec.astype(np.float32)

    def embed_batch(self, texts: List[str]) -> np.ndarray:
        """
        Generates a 2D normalized float32 NumPy matrix of shape (N, dimension) for N texts.
        """
        if not texts:
            return np.empty((0, self.dimension), dtype=np.float32)
        return np.vstack([self.embed_text(t) for t in texts]).astype(np.float32)

    def cosine_similarity(self, vec1: np.ndarray, vec2: np.ndarray) -> float:
        """Computes scalar cosine similarity between two 1D normalized embedding vectors."""
        v1 = vec1.astype(np.float32)
        v2 = vec2.astype(np.float32)
        norm1 = np.linalg.norm(v1)
        norm2 = np.linalg.norm(v2)
        if norm1 > 1e-12:
            v1 = v1 / norm1
        if norm2 > 1e-12:
            v2 = v2 / norm2
        return float(np.dot(v1, v2))

    def batch_cosine_similarity(self, query_vec: np.ndarray, doc_matrix: np.ndarray) -> np.ndarray:
        """
        Vectorized computation of cosine similarities between a 1D query vector (dimension,)
        and a 2D matrix of document vectors (N, dimension). Returns 1D array of scores (N,).
        """
        if doc_matrix.size == 0:
            return np.empty(0, dtype=np.float32)
        q = query_vec.astype(np.float32)
        q_norm = np.linalg.norm(q)
        if q_norm > 1e-12:
            q = q / q_norm
        return np.dot(doc_matrix.astype(np.float32), q)

File: ocaml/lib/embeddings.ml
(**
   embeddings.ml - Deterministic 256-D Offline Feature Hashing Embedder.
   Uses multi-scale signed feature hashing with subword n-grams and domain token boosting.
   100% offline, zero external models, zero cloud APIs, zero mock dependencies.
*)

let default_dimension = 256

(** Precomputed IEEE 802.3 CRC32 lookup table (polynomial 0xEDB88320). *)
let crc32_table =
  let poly = 0xEDB88320 in
  Array.init 256 (fun i ->
    let crc = ref i in
    for _ = 0 to 7 do
      if (!crc land 1) <> 0 then
        crc := ((!crc lsr 1) lxor poly) land 0xFFFFFFFF
      else
        crc := (!crc lsr 1) land 0xFFFFFFFF
    done;
    !crc
  )

let crc32 (s : string) : int =
  let crc = ref 0xFFFFFFFF in
  for i = 0 to String.length s - 1 do
    let b = Char.code s.[i] in
    let idx = (!crc lxor b) land 0xFF in
    crc := ((!crc lsr 8) lxor crc32_table.(idx)) land 0xFFFFFFFF
  done;
  (!crc lxor 0xFFFFFFFF) land 0xFFFFFFFF

let sign_of_token (s : string) : float =
  let digest = Digest.string s in
  if (Char.code digest.[0]) land 1 = 0 then 1.0 else -1.0

let is_word_boundary_char (c : char) : bool =
  not ((c >= 'a' && c <= 'z') || (c >= '0' && c <= '9') || c = '_')

let is_token_char (c : char) : bool =
  (c >= 'a' && c <= 'z') || (c >= '0' && c <= '9') ||
  c = '_' || c = '-' || c = '.' || c = ':' || c = '#' ||
  c = '[' || c = ']' || c = '=' || c = '"'

let tokenize (text : string) : (string * float) list =
  let trimmed = String.trim text in
  if String.length trimmed = 0 then
    [("__EMPTY__", 1.0)]
  else
    let cleaned = String.lowercase_ascii trimmed in
    let len = String.length cleaned in

    (* 1. Extract status codes (4xx and 5xx) *)
    let status_tokens = ref [] in
    for i = 0 to len - 3 do
      let c0 = cleaned.[i] in
      let c1 = cleaned.[i + 1] in
      let c2 = cleaned.[i + 2] in
      let is_status_start = (c0 = '4' || c0 = '5') &&
                            (c1 >= '0' && c1 <= '9') &&
                            (c2 >= '0' && c2 <= '9') in
      if is_status_start then
        let left_bound = (i = 0 || is_word_boundary_char cleaned.[i - 1]) in
        let right_bound = (i + 3 = len || is_word_boundary_char cleaned.[i + 3]) in
        if left_bound && right_bound then
          let code = String.sub cleaned i 3 in
          status_tokens := ("status:" ^ code, 3.0) :: !status_tokens
    done;
    let status_tokens = List.rev !status_tokens in

    (* 2. Extract lexical words and domain components *)
    let words = ref [] in
    let i = ref 0 in
    while !i < len do
      while !i < len && not (is_token_char cleaned.[!i]) do
        incr i
      done;
      let start_pos = !i in
      while !i < len && is_token_char cleaned.[!i] do
        incr i
      done;
      if !i > start_pos then
        let w = String.sub cleaned start_pos (!i - start_pos) in
        words := w :: !words
    done;
    let word_list = List.rev !words in
    let word_tokens = List.map (fun w -> ("w:" ^ w, 1.5)) word_list in

    (* 3. Extract word bigrams *)
    let bigram_tokens = ref [] in
    let rec make_bigrams = function
      | w1 :: (w2 :: _ as rest) ->
          bigram_tokens := ("bi:" ^ w1 ^ "_" ^ w2, 2.0) :: !bigram_tokens;
          make_bigrams rest
      | _ -> ()
    in
    make_bigrams word_list;
    let bigram_tokens = List.rev !bigram_tokens in

    (* 4. Extract character 3-grams and 4-grams *)
    let ngram_tokens = ref [] in
    List.iter (fun w ->
      let w_len = String.length w in
      if w_len >= 3 then
        for j = 0 to w_len - 3 do
          ngram_tokens := ("3g:" ^ String.sub w j 3, 0.5) :: !ngram_tokens
        done;
      if w_len >= 4 then
        for j = 0 to w_len - 4 do
          ngram_tokens := ("4g:" ^ String.sub w j 4, 0.5) :: !ngram_tokens
        done
    ) word_list;
    let ngram_tokens = List.rev !ngram_tokens in

    status_tokens @ word_tokens @ bigram_tokens @ ngram_tokens

let l2_norm (vec : float array) : float =
  let sum = ref 0.0 in
  for i = 0 to Array.length vec - 1 do
    sum := !sum +. (vec.(i) *. vec.(i))
  done;
  sqrt !sum

let embed_text ?(dimension = default_dimension) (text : string) : float array =
  if dimension <= 0 then invalid_arg "Embedding dimension must be positive";
  let vec = Array.make dimension 0.0 in
  let tokens = tokenize text in
  List.iter (fun (tok, weight) ->
    let c = crc32 tok in
    let idx = c mod dimension in
    let sign = sign_of_token tok in
    vec.(idx) <- vec.(idx) +. (sign *. weight)
  ) tokens;
  let norm = l2_norm vec in
  if norm > 1e-12 then
    Array.map (fun x -> x /. norm) vec
  else (
    let out = Array.make dimension 0.0 in
    out.(0) <- 1.0;
    out
  )

let embed_batch ?(dimension = default_dimension) (texts : string list) : float array list =
  List.map (embed_text ~dimension) texts

let cosine_similarity (v1 : float array) (v2 : float array) : float =
  let len1 = Array.length v1 in
  let len2 = Array.length v2 in
  if len1 = 0 || len1 <> len2 then 0.0
  else
    let n1 = l2_norm v1 in
    let n2 = l2_norm v2 in
    if n1 <= 1e-12 || n2 <= 1e-12 then 0.0
    else
      let dot = ref 0.0 in
      for i = 0 to len1 - 1 do
        dot := !dot +. (v1.(i) *. v2.(i))
      done;
      let sim = !dot /. (n1 *. n2) in
      if sim > 1.0 then 1.0
      else if sim < -1.0 then -1.0
      else sim

let batch_cosine_similarity (q : float array) (docs : float array list) : float list =
  List.map (fun doc -> cosine_similarity q doc) docs
